optima/geospatial.py

```diff
@@ -0,0 +1,335 @@
+                sys.stdout = oldstdout  # The important bit.
+                event.accept()
+            else:
+                event.ignore()    
+    
+    ## Housekeeping
+    fig = figure(); close(fig) # Open and close figure...dumb, no? Otherwise get "QWidget: Must construct a QApplication before a QPaintDevice"
+    geoguiwindow = GAWidget() # Create panel widget
+    geoguiwindow.setGeometry(100, 100, wid, hei)
+    geoguiwindow.setWindowTitle('Optima geospatial analysis')
+    projectslist = []
+    
+    ##############################################################################################################################
+    ## Define functions
+    ##############################################################################################################################
+
+    
+    def _loadproj():
+        ''' Little helper function to load a project, since used more than once '''
+        filepath = QtGui.QFileDialog.getOpenFileNames(caption='Choose project file', filter='*'+projext)
+        project = None
+        try: 
+            project = loadobj(filepath, verbose=0)
+        except: 
+            print('Could not load file "%s"' % filepath)
+            return None
+        try: 
+            assert type(project)==Project
+            return project
+        except: 
+            print('File "%s" is not an Optima project file' % filepath)
+            return None
+        
+        
+    def makesheet():
+        ''' Create a geospatial spreadsheet template based on a project file '''
+        
+        ## 1. Load a project file
+        project = _loadproj()
+            
+        ## 2. Get destination filename
+        spreadsheetpath = QtGui.QFileDialog.getSaveFileName(caption='Save geospatial spreadsheet file', filter='*.xlsx')
+        
+        ## 3. Extract data needed from project (population names, program names...)
+        # ...
+        
+        ## 4. Generate and save spreadsheet
+        # ...
+
+        return None
+        
+    
+    def makeproj():
+        ''' Create a series of project files based on a seed file and a geospatial spreadsheet '''
+        
+        ## 1. Load a project file -- WARNING, could be combined with the above!
+        project = _loadproj()
+        
+        ## 2. Load a spreadsheet file
+        spreadsheetpath = QtGui.QFileDialog.getOpenFileNames(caption='Choose geospatial spreadsheet', filter='*.xlsx')
+        
+        ## 3. Get a destination folder
+        destination = QtGui.QFileDialog.getExistingDirectory(caption='Choose output folder')
+        
+        ## 4. Read the spreadsheet
+        # ...
+        
+        ## 5. Calibrate each project file according to the data entered for it in the spreadsheet
+        # ...
+        
+        ## 6. Save each project file into the directory
+        # ...
+        
+        return None
+
+
+    def create():
+        ''' Create a portfolio by selecting a list of projects; silently skip files that fail '''
```

Comment:
so if the user is selecting 30 files and accidentally gets a wrong one, it doesn't kill everything; the user can add missing ones in later...